Count store support for cartesian products

pontusmelke opened the pull request:
For queries such as `MATCH (m),(n),(o),... RETURN count(*)` there is no
need to do the full cartesian product, we can use the count store
directly instead.

pontusmelke committed:
Count store support for cartesian products

For queries such as `MATCH (m),(n),(o),... RETURN count(*)` there is no
need to do the full cartesian product, we can use the count store
directly instead.

craigtaverner commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala:
```diff
@@ -43,6 +43,12 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
     result.columnAs[Node]("n").toList should equal(List(n1, n2, n3, n4, n5))
   }
 
+  test("foo") {
```
Test name?

craigtaverner replied:
Perhaps this was temporary code and should be deleted. This looks superseded by the tests in MatchAggregationsBackedByCountStoreAcceptanceTest.scala.

pontusmelke replied:
yes you are right, I meant to delete this one...

craigtaverner commented on community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/LogicalPlan2PlanDescription.scala:
```diff
@@ -83,7 +83,7 @@ case class LogicalPlan2PlanDescription(idMap: Map[LogicalPlan, Id], readOnly: Bo
 
       case NodeCountFromCountStore(IdName(variable), labelName, _) =>
-        val arguments = Seq(CountNodesExpression(variable, labelName.map(_.name)))
+        val arguments = Seq(CountNodesExpression(variable, labelName.map(l => l.map(_.name))))
```
Would be better to pluralize the variable name

craigtaverner commented on community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planDescription/InternalPlanDescription.scala:
```diff
@@ -116,7 +116,7 @@ object InternalPlanDescription {
     case class ExpandExpression(from: String, relName: String, relTypes:Seq[String], to: String,
                                 direction: SemanticDirection, minLength: Int, maxLength: Option[Int]) extends Argument
-    case class CountNodesExpression(ident: String, label: Option[String]) extends Argument
+    case class CountNodesExpression(ident: String, label: List[Option[String]]) extends Argument
```
Would be better to pluralize the parameter name.

pontusmelke committed:
Removed bad test and pluralize

After code review:
- Remove a test that was not meant to be there
- Pluralize names

craigtaverner commented on community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/NodeCountFromCountStore.scala:
```diff
@@ -23,7 +23,7 @@ import org.neo4j.cypher.internal.frontend.v3_3.ast.LabelName
 
 
-case class NodeCountFromCountStore(idName: IdName, labelName: Option[LabelName], argumentIds: Set[IdName])
+case class NodeCountFromCountStore(idName: IdName, labelName: List[Option[LabelName]], argumentIds: Set[IdName])
```
Perhaps pluralize labelNames
